### DISCUSSION TITLE
Is there a method that gets called before any query/mutation that we can override for logging purposes?

### DISCUSSION THREAD

**Discussion by drewgallagher** [rebing/graphql-laravel#1096] [Q&A]
Right now in our project we have the Log facade calls in every mutation/query, however this is a lot of code to maintain and refactor if needed. 

We have hundreds of mutations/queries. 

Is there a method built in we can override that gets called for every mutation/query, that we can add a log statement in one spot. 

Something like a before() function where we could log the query/mutation that is getting called with the given request body? 

This would cut our Log calls and imports from N files (based on the query/mutation count) to 1

**Comment by mfn** [comment]
With the latest major version, a `Query` (and `Mutation`, for that matter) can have multiple `$middleware` which can do this. They must implement a `handle` method and receive, see https://github.com/rebing/graphql-laravel#resolver-middleware for more.

That said, I implemented what you said but did it on the http level, via regular Laravel middlewares on the graphql http route (in `app/Http/Kernel.php`). In my case this is sufficient enough, more comaptible, less complex; but surely depends how rich you need the information. In my case, just logging the literal "graphql query string" as received was enough, i.e. the incoming JSON payload.

**Reply by drewgallagher** [reply]
Ok cool. are you able to provide a code example of that logging functionality in the Kernel that I can reference? Thanks!

**Reply by mfn** [reply]
It's nothing fancy and not really specific to GraphQL and you could use it to log out kind of http requests; this just pseudo-code how to structure it:
```php
<?php
class LoggingMiddleware
{
    private ?Log $log = null;

    /**
     * @return mixed
     */
    public function handle(Request $request, Closure $next)
    {
        $this->log = new Log();

        return $next($request);
    }

    public function terminate(Request $request, Response $response): void
    {
        if (!$this->log) {
            return;
        }

        $this->log->update($request, $response);
        $this->log->save();
    }
}
```

Of course the "meat" happens in `->update()`, it receives both objects and extracts whatever information it wants to log

This is just one way of doing it and not necessarily the best, dunno
HTH
